server/src/main/java/org/elasticsearch/action/search/SearchResponseMerger.java

```diff
@@ -65,35 +71,78 @@
+
     final int from;
     final int size;
     final int trackTotalHitsUpTo;
     private final SearchTimeProvider searchTimeProvider;
     private final AggregationReduceContext.Builder aggReduceContextBuilder;
     private final List<SearchResponse> searchResponses = new CopyOnWriteArrayList<>();
+    private final SearchResponse.Clusters clusters;
+    private final SearchProgressListener progressListener;
+    private final int numExpectedResponses;
+
+    SearchResponseMerger(
+        int from,
+        int size,
+        int trackTotalHitsUpTo,
+        SearchTimeProvider searchTimeProvider,
+        AggregationReduceContext.Builder aggReduceContextBuilder,
+        SearchResponse.Clusters clusters,
+        SearchProgressListener progressListener
+    ) {
+        this(from, size, trackTotalHitsUpTo, searchTimeProvider, aggReduceContextBuilder, clusters, progressListener, -1);
+    }
 
+    /**
+     *
+     * @param from
+     * @param size
+     * @param trackTotalHitsUpTo
+     * @param searchTimeProvider
+     * @param aggReduceContextBuilder
+     * @param clusters
+     * @param progressListener
+     * @param numTotalClusters Specify the total number of clusters to turn on an optimization where an incremental
+     *                         merge will not be done in the add method for the last search response to come in
+     *                         (since the user of the Merger object is going to call {getMergedResponse} after
+     *                         all have been added.
+     */
     SearchResponseMerger(
         int from,
         int size,
         int trackTotalHitsUpTo,
         SearchTimeProvider searchTimeProvider,
-        AggregationReduceContext.Builder aggReduceContextBuilder
+        AggregationReduceContext.Builder aggReduceContextBuilder,
+        SearchResponse.Clusters clusters,
+        SearchProgressListener progressListener,
+        int numTotalClusters
     ) {
         this.from = from;
         this.size = size;
         this.trackTotalHitsUpTo = trackTotalHitsUpTo;
         this.searchTimeProvider = Objects.requireNonNull(searchTimeProvider);
         this.aggReduceContextBuilder = aggReduceContextBuilder; // might be null if there are no aggregations
+        this.clusters = clusters;
+        this.progressListener = progressListener;
+        this.numExpectedResponses = numTotalClusters;
     }
 
     /**
      * Add a search response to the list of responses to be merged together into one.
-     * Merges currently happen at once when all responses are available and {@link #getMergedResponse(Clusters)} )} is called.
+     * Merges currently happen at once when all responses are available and {@link #getMergedResponse()} )} is called.
      * That may change in the future as it's possible to introduce incremental merges as responses come in if necessary.
      */
-    void add(SearchResponse searchResponse) {
+    synchronized void add(SearchResponse searchResponse) {
```

Comment:
Now that getMergedResponse can be called from multiple threads (not just the callback thread of the final cluster to report back), I've synchonized this method and the getMergedResponse method to avoid consistency errors.